src/hip_module.cpp

```diff
@@ -480,6 +480,116 @@ hipError_t hipLaunchCooperativeKernel(const void* f, dim3 gridDim,
+    }
+
+    // prepare all kernel descriptors for initializing the GWS and the main kernels per device
+    for (int i = 0; i < numDevices; ++i) {
+        const hipLaunchParams& lp = launchParamsList[i];
+        if (lp.stream == nullptr) {
+            free(gwsKds);
+            free(kds);
+            return ihipLogStatus(hipErrorNotInitialized);
+        }
+
+        gwsKds[i] = hip_impl::get_program_state().kernel_descriptor(reinterpret_cast<std::uintptr_t>(&init_gws),
+                hip_impl::target_agent(lp.stream));
+        if (gwsKds[i] == nullptr) {
+            free(gwsKds);
+            free(kds);
+            return ihipLogStatus(hipErrorInvalidValue);
+        }
+        hip_impl::kernargs_size_align gwsKargs = hip_impl::get_program_state().get_kernargs_size_align(
+                reinterpret_cast<std::uintptr_t>(&init_gws));
+        gwsKds[i]->_kernarg_layout = *reinterpret_cast<const std::vector<std::pair<std::size_t, std::size_t>>*>(
+                gwsKargs.getHandle());
+
+
+        kds[i] = hip_impl::get_program_state().kernel_descriptor(reinterpret_cast<std::uintptr_t>(lp.func),
+                hip_impl::target_agent(lp.stream));
+        if (kds[i] == nullptr) {
+            free(gwsKds);
+            free(kds);
+            return ihipLogStatus(hipErrorInvalidValue);
+        }
+        hip_impl::kernargs_size_align kargs = hip_impl::get_program_state().get_kernargs_size_align(
+                reinterpret_cast<std::uintptr_t>(lp.func));
+        kds[i]->_kernarg_layout = *reinterpret_cast<const std::vector<std::pair<std::size_t, std::size_t>>*>(
+                kargs.getHandle());
+    }
+
+    // lock all streams before launching the blit kernels for initializing the GWS and main kernels to each device
+    for (int i = 0; i < numDevices; ++i) {
+        LockedAccessor_StreamCrit_t streamCrit(launchParamsList[i].stream->criticalData(), false);
+#if (__hcc_workweek__ >= 19213)
+        streamCrit->_av.acquire_locked_hsa_queue();
+#endif
+    }
+
+    // launch the init_gws kernel to initialize the GWS followed by launching the main kernels for each device
+    for (int i = 0; i < numDevices; ++i) {
+        const hipLaunchParams& lp = launchParamsList[i];
+
+        void *gwsKernelParam[1];
+        uint nwm1 = (lp.gridDim.x * lp.gridDim.y * lp.gridDim.z) - 1;
+        gwsKernelParam[0] = &nwm1;
+
+        result = ihipModuleLaunchKernel(tls, gwsKds[i], 1, 1, 1, 1, 1, 1,
+                0, lp.stream, gwsKernelParam, nullptr, nullptr, nullptr, 0, true);
+
+        if (result != hipSuccess) {
+            for (int j = 0; j < numDevices; ++j) {
+                launchParamsList[j].stream->criticalData().unlock();
+#if (__hcc_workweek__ >= 19213)
+                launchParamsList[j].stream->criticalData()._av.release_locked_hsa_queue();
+#endif
+            }
+            return ihipLogStatus(hipErrorLaunchFailure);
+        }
+
+        result = ihipModuleLaunchKernel(tls, kds[i],
```

Comment:
What happens if gwsKds is successful on device0, device1,.. but fails on deviceN? We would end up enqueuing the actual kernels on some devices before returning LaunchFailure. Is that fine?

Reply:
I think this scenario is less likely since the checks for gwsKds is done before the launch and even if it happens for whatever reason, not much can be done for already enqueued main kernels and only an error is returned to the caller notifying the launch failure for the API.
We could also have similar issue if the gwsKds is successful on all devices but main kernel is successful only on device0, device1, .. but fails on deviceN. in this case a launch failure error is similarly reported to the caller and the API call still fails..